index.js

```diff
@@ -51,21 +38,33 @@ export function createHandler (functions) {
   return function handler (self) {
     const cache = {}
 
-    self.addEventListener('message', function (event) {
+    self.addEventListener('message', async function (event) {
       const {command, id, message} = event.data
-      Promise
-        .resolve(functions[command].call(null, cache, message))
-        .then(results => {
-          self.postMessage({
-            command,
-            id,
-            message: results
-          })
-        })
-        .catch(e => {
-          e.id = id
-          throw e
+
+      let results
+      try {
+        results = await functions[command].call(null, cache, message)
+      } catch (error) {
+        // nonstandard but will just be undefined on browsers that don't support
+        const { fileName, lineNumber } = error
+        self.postMessage({
+          command,
+          id,
+          error: {
+            filename: fileName,
+            lineno: lineNumber,
+            message: error.message
+          }
         })
+
+        return
```

Comment:
Instead of returning early I would just hoist the `self.postMessage` below into the try block. But no big deal.